File: university-service/src/test/java/ua/com/foxminded/courseproject/repository/TeacherRepositoryTest.java
package ua.com.foxminded.courseproject.repository;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.data.util.Streamable;
import org.springframework.test.context.jdbc.Sql;
import org.springframework.transaction.annotation.Transactional;
import ua.com.foxminded.courseproject.entity.Teacher;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;

@SpringBootTest
@Sql(value = "classpath:initial_data.sql")
@Transactional
class TeacherRepositoryTest {

    @Autowired
    private TeacherRepository repository;

    @Test
    void findAll_shouldReturnListTeachers() {
        Integer expectedSize = 2;

        List<Teacher> teachers = Streamable.of(repository.findAll()).toList();

        assertEquals(expectedSize, teachers.size());
    }

    @Test
    void findById_shouldReturnTeacher_whenIdExists() {
        UUID id = UUID.fromString("e966f608-4621-11ed-b878-0242ac120002");
        String expectedFirstname = "Yulia";

        Teacher actual = repository.findById(id).get();

        assertEquals(expectedFirstname, actual.getFirstName());
    }

    @Test
    void findById_shouldReturnEmptyOptional_whenIdNotExists() {
        UUID id = UUID.fromString("e966f7c0-4621-11ed-b838-0242ac120002");

        Optional<Teacher> actual = repository.findById(id);

        assertEquals(true, actual.isEmpty());
    }

    @Test
    void save_shouldAddTeacher_whenTeacherNotExists() {
        String testStr = "test";
        Teacher expected = new Teacher();
        expected.setFirstName(testStr);
        expected.setLastName(testStr);
        expected.setDegree(testStr);
        expected.setFirstDay(LocalDate.now());
        expected.setBirthDay(LocalDate.now());
        expected.setRank(testStr);
        expected.setTitle(testStr);
        expected.setSalary(1111);

        repository.save(expected);

        assertEquals(expected, repository.findById(expected.getId()).get());
    }

    @Test
    void save_shouldChangeTeacher_whenTeacherExists() {
        Teacher expected = Streamable.of(repository.findAll()).stream().findAny().get();
        expected.setLastName("test");

        repository.save(expected);

        assertEquals(expected, repository.findById(expected.getId()).get());
    }

    @Test
    void delete_shouldDeleteTeacher_whenTeacherExists() {
        Teacher teacher = Streamable.of(repository.findAll()).stream().findAny().get();
        Integer expectedSize = 1;

        repository.delete(teacher);
        List<Teacher> teachers = Streamable.of(repository.findAll()).stream().toList();

        assertEquals(expectedSize, teachers.size());
    }

    @Test
    void delete_shouldNotDeleteTeachers_whenTeacherNotExists() {
        String testStr = "test";
        Teacher teacher = new Teacher();
        teacher.setFirstName(testStr);
        teacher.setLastName(testStr);
        teacher.setDegree(testStr);
        teacher.setFirstDay(LocalDate.now());
        teacher.setBirthDay(LocalDate.now());
        teacher.setRank(testStr);
        teacher.setTitle(testStr);
        teacher.setSalary(1111);
        Integer expectedSize = 2;

        repository.delete(teacher);
        List<Teacher> teachers = Streamable.of(repository.findAll()).toList();

        assertEquals(expectedSize, teachers.size());
    }

}

File: university-service/src/test/kotlin/ua/com/foxminded/courseproject/repository/TeacherRepositoryTest.kt
package ua.com.foxminded.courseproject.repository

import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.data.mongo.DataMongoTest
import reactor.test.StepVerifier
import ua.com.foxminded.courseproject.config.DBTestConfig
import ua.com.foxminded.courseproject.entity.Teacher
import java.time.LocalDate
import java.util.*


@DataMongoTest
internal open class TeacherRepositoryTest : DBTestConfig() {

    @Autowired
    private lateinit var repository: TeacherRepository

    @Test
    fun findAll_shouldReturnFluxTeachers() {
        val expectedSize = 2

        val teachers = repository.findAll()

        StepVerifier.create(teachers)
            .expectNextCount(expectedSize.toLong())
            .verifyComplete()


    }

    @Test
    fun findById_shouldReturnTeacher_whenIdExists() {
        val id = UUID.fromString("e966f608-4621-11ed-b878-0242ac120002")
        val expectedFirstname = "Yulia"

        val actual = repository.findById(id)

        StepVerifier.create(actual)
            .expectNextMatches { expectedFirstname == it.firstName }
            .verifyComplete()
    }

    @Test
    fun findById_shouldReturnEmptyMono_whenIdNotExists() {
        val id = UUID.fromString("e966f7c0-4621-11ed-b838-0242ac120002")

        val actual = repository.findById(id)

        StepVerifier.create(actual)
            .expectNextCount(0)
            .verifyComplete()
    }

    @Test
    fun save_shouldAddTeacher_whenTeacherNotExists() {
        val testStr = "test"
        val expected = Teacher()
        expected.firstName = testStr
        expected.lastName = testStr
        expected.degree = testStr
        expected.firstDay = LocalDate.now()
        expected.birthDay = LocalDate.now()
        expected.rank = testStr
        expected.title = testStr
        expected.salary = 1111

        repository.save(expected).block()

        expected.id?.let { repository.findById(it) }?.let {
            StepVerifier.create(it)
                .expectNextMatches { it.equals(expected) }
                .verifyComplete()
        }
    }

    @Test
    fun save_shouldChangeTeacher_whenTeacherExists() {
        val expected = repository.findAll().blockFirst()
        expected.lastName = "test"

        repository.save(expected).block()

        expected.id?.let { repository.findById(it) }?.let {
            StepVerifier.create(it)
                .expectNextMatches { it.equals(expected) }
                .verifyComplete()
        }
    }

    @Test
    fun delete_shouldDeleteTeacher_whenTeacherExists() {
        val teacher = repository.findAll().blockFirst()
        val expectedSize = 1

        repository.delete(teacher).block()

        StepVerifier.create(repository.findAll())
            .expectNextCount(expectedSize.toLong())
            .verifyComplete()
    }

    @Test
    open fun delete_shouldNotDeleteTeachers_whenTeacherNotExists() {
        val testStr = "test"
        val teacher = Teacher()
        teacher.firstName = testStr
        teacher.lastName = testStr
        teacher.degree = testStr
        teacher.firstDay = LocalDate.now()
        teacher.birthDay = LocalDate.now()
        teacher.rank = testStr
        teacher.title = testStr
        teacher.salary = 1111
        val expectedSize = 2

        repository.delete(teacher).block()

        StepVerifier.create(repository.findAll())
            .expectNextCount(expectedSize.toLong())
            .verifyComplete()
    }
}
